File: snap.R
# Take a normal pair/permutation generating function (assumed to ignore jokers when matching cards up),
# and add pair contributions from the jokers. This will consequently reduce the number of normal pairs
# to otherwise find, and increase the number of cards consumed.
with.joker.pairs <- function(fn) {
	
  all.joker.blocks <- function(jokers) {
    if (sum(jokers) > 2)
      c(list(jokers), all.joker.blocks(next.blocks(jokers)))
    else if (sum(jokers) == 2)
      c(list(2), list(0))
    else
      list(0)
  }
  
  joker.blocks.allowed <- function(ranks, suits, pairs, cards) {
    function(j) {
      joker.cards <- sum(j)
      joker.pairs <- if (joker.cards > 0) sum(j-1) else 0
      joker.pairs <= pairs && joker.cards <= cards && (pairs-joker.pairs) <= floor((suits-1)*min(ranks*suits, cards-joker.cards)/suits)
    }
  }

  function(ranks, suits, jokers, pairs, cards, blocks) {
    if (jokers < 2)
      fn(ranks, suits, jokers, pairs, cards, blocks) # cannot make joker pairs
    else {
      joker.blocks <- Filter(joker.blocks.allowed(ranks, suits, pairs, cards), all.joker.blocks(jokers))
      results <- lapply(joker.blocks, function(j) {
        jokers.used <- sum(j)
        joker.pairs <- if (jokers.used > 0) sum(j-1) else 0
        joker.perms <- block.permutations(1, jokers, jokers.used > 0, j)
        results <- lapply(fn(ranks, suits, jokers-jokers.used, pairs-joker.pairs, cards-jokers.used, blocks), function(x) {
          c(x[1] * joker.perms, x[2] + jokers.used)
        })
      })
      unlist(results, recursive=FALSE)
    }
  }
}


block.permutations <- function(ranks, suits, ranks.used, blocks) {
  if (ranks.used == 0)
    1
  else {
    f <- function(acc, n) {
      existing.perms <- acc[1]
      suit.cards.remaining <- acc[2]
      new.perms <- prod((suit.cards.remaining-n+1):suit.cards.remaining)
      c(existing.perms * new.perms, suit.cards.remaining-n)
    }
    degeneracy <- prod(sapply(table(blocks), factorial))
    choose(ranks, ranks.used) * (Reduce(f, blocks, c(1, suits))[1] / degeneracy)^ranks.used
  }
}


next.blocks <- function(blocks) {
  total <- sum(blocks)
  bigger.than.two = Filter(function(n) n>2, blocks)
  # If total cards < 4, we can't split into separate groups of pairs, already atomic.
  # If no cards are > 2, or only one is and it's 3, we cannot split further.
  # If we cannot split further, jump down to the next group size.
  if (total < 4 || length(bigger.than.two) == 0 || (length(bigger.than.two) == 1 && bigger.than.two[1] == 3))
    total - 1
  else {
    # Strategy: try and split into 2 halves first, then reduce size of first group until it terminates at 2.
    # Apply this policy recursively to generate successive blocks.
    spread <- function(blocks) {
      total <- sum(blocks)
      count <- length(blocks)
      if (count == 1)
        c(floor(total/2), total - floor(total/2))
      else if (blocks[1] == 2)
        c(2, spread(blocks[2:count]))
      else {
        blocks[1] <- blocks[1] - 1
        blocks[2] <- blocks[2] + 1
        blocks
      }
    }
    spread(blocks)
  }
}


permutations <- function(ranks, suits, jokers, pairs, cards, blocks) {
  block.cards <- sum(blocks)
  if (block.cards == 1)
    list(c(1,0))
  else {
    block.pairs <- sum(blocks-1)
    min.blocks <- max(0, pairs-ranks*(block.cards-2))
    max.blocks <- ranks
    n = Filter(function(n) (pairs >= n*block.pairs && n*block.cards <= cards), min.blocks:max.blocks)
    results <- lapply(n, function(n) {
      pairs.created <- n*block.pairs
      cards.used <- n*block.cards
      others <- permutations(ranks-n, suits, jokers, pairs-pairs.created, cards-cards.used, next.blocks(blocks))
      lapply(others, function(x) c(x[1] * block.permutations(ranks, suits, n, blocks), x[2] + cards.used))
    })
    unlist(results, recursive=FALSE)
  }
}


#
#  Calculate the probability of a pair occurring in the first k cards
#  k - the depth
#
pair.probability <- function(k, ranks, suits, jokers=0) {
  deck <- ranks*suits + jokers
  at.least.p.pairs <- function(p) {
    probability <- function(tuple) {
      pair.permutations <- tuple[1]
      cards.used <- tuple[2]
      pair.permutations * prod((k-cards.used+1):(k-p)) / prod((deck-cards.used+1):deck)
    }
    fn <- with.joker.pairs(permutations)
    perms <- fn(ranks, suits, jokers, p, k, blocks=min(suits, k))
    sum(sapply(perms, probability))
  }

  # Use the inclusion-exclusion principle to calculate the probability of a pair existing
  max.pairs <- floor((suits-1)*k/suits)
  if (max.pairs > 0) sum(sapply(1:max.pairs, function(p) ((-1)^(p+1) * at.least.p.pairs(p)))) else 0
}


#
#  Calculate the probability of the first pair occurring at k
#  k - the location of the first pair
#
first.pair.probability <- function(k, ranks, suits) {
  if (k < 2)
    0
  else {
    # m is the number of cards making up the given pair at k (ie 2 if first pair in pack)
    # When called recursively, this function returns the probability that a pair precedes all the
    # pairs in a block of m cards of the same suit, where the block terminates at k.
    probability.of.pair.before <- function(k, ranks, suits, m=2) {
      if (k <= m)
        0
      else if (m == suits)
        pair.probability(k-suits, ranks-1, suits)
      else
        pair.probability(k-m, ranks-1, suits, suits-m) +
          (1-probability.of.pair.before(k, ranks, suits, m+1))*(suits-m)/(ranks*suits-m)
    }
    (1-probability.of.pair.before(k, ranks, suits))*(suits-1)/(ranks*suits-1)
  }
}


#
#  Calculate the mean location of the first pair
#
first.pair.mean.location <- function(ranks, suits) {
  k <- 1:(ranks*suits)
  return (sum(k * sapply(k, function(k) first.pair.probability(k, ranks, suits))))
}


#
#  Calculate the mean location of the first pair, using all available cores
#
first.pair.mean.location.parallel <- function(ranks, suits) {
  k <- 1:(ranks*suits)
  cores <- parallel::detectCores()
  fun <- function(k) {
    print(sprintf("%s %s", date(), k))
    first.pair.probability(k, ranks, suits)
  }
  results <- unlist(parallel::mclapply(k, fun, mc.cores=cores))
  sum(k * results)
}

File: Snap.scala
package snap

import scala.util.Random

object Snap extends App with Utils {

  /**
   * Example: r=13, s=4 (standard pack of playing cards)
   * We have a block of 4 matching cards x 2: ranksUsed = 2 blocks = List(4)
   * We have two single pairs for a single rank: ranksUsed = 1, blocks = List(2,2)
   * We have a single pair x 4: ranksUsed = 4, blocks = List(2)
   *
   * This method returns the number of permuations (where the order of the blocks is unimportant)
   */
  def blockPermutations(ranks: Int, suits: Int, ranksUsed: Int, blocks: List[Int]): BigInt = {
    if (blocks.sum > suits) throw new IllegalArgumentException("The total number of cards per rank cannot exceed the number of suits")
    if (ranksUsed == 0)
      1
    else {
      val (_, perms) = blocks.foldLeft((suits, 1)) {
        (cards, block) =>
          cards match {
            case (cardsRemaining, existingPerms) =>
              val newPerms = ((cardsRemaining - block + 1) to cardsRemaining).product
              (cardsRemaining - block, existingPerms * newPerms)
          }
      }
      val degeneracy = (blocks.groupBy(block => block) map { case (_, values) => factorial(values.length) }).product
      binomialCoefficient(ranks, ranksUsed) * (perms / degeneracy).pow(ranksUsed)
    }
  }

  /**
   * Return the next set of blocks to consider for a suit.
   * Best demonstrated with examples:
   * List(4) -> List(2,2)
   * List(2,2) -> List(3)
   * List(3) -> List(2)
   *
   * Expected usage: start with List(suits). Then this method returns the next configuration of cards
   * to consider. We move from maximum number of cards available -> 2 for a single pair
   */
  def nextBlocks(blocks: List[Int]): List[Int] = {
    val total = blocks.sum
    if (total < 2) throw new IllegalArgumentException("Cards cannot be broken down further")
    val biggerThanTwo = blocks filter (_ > 2)
    if (total < 4 || biggerThanTwo.isEmpty || biggerThanTwo == Seq(3))
      List(total - 1)
    else {
      def spread(blocks: List[Int]): List[Int] = {
        (blocks: @unchecked) match {
          case block :: Nil =>
            val total = blocks.sum
            val half = math.floor(total / 2).toInt
            List(half, total - half)
          case 2 :: tail =>
            2 :: spread(tail)
          case first :: second :: tail =>
            (first - 1) :: (second + 1) :: tail
        }
      }
      spread(blocks)
    }
  }

  /**
   * Permutations
   */
  def permutations(ranks: Int, suits: Int, jokers: Int, pairs: Int, cards: Int, blocks: List[Int]): List[(BigInt, Int)] = {
    if (blocks == List(1))
      List((1, 0))
    else {
      val blockCards = blocks.sum
      val blockPairs = (blocks map (_ - 1)).sum
      val minBlocks = math.max(0, pairs - ranks * (blockCards - 2)) // deduct max number of pairs available for next block down..
      val maxBlocks = ranks
      (minBlocks to maxBlocks).toList filter (n => pairs >= n * blockPairs && n * blockCards <= cards) flatMap {
        n =>
          // Use n * blocks
          // Then recursively find remaining pairs (if any) using other block configurations, remembering that
          // the number of remaining ranks is reduced by n, with the pairs required/cards used also suitably updated
          val pairsCreated = n * blockPairs
          val cardsUsed = n * blockCards
          val others = permutations(ranks - n, suits, jokers, pairs - pairsCreated, cards - cardsUsed, nextBlocks(blocks))
          val blockPerms = blockPermutations(ranks, suits, n, blocks)
          others map { case (x, y) => (x * blockPerms, y + cardsUsed) }
      }
    }
  }

  /**
   * Incorporate jokers
   */
  def withJokers(permutations: (Int, Int, Int, Int, Int, List[Int]) => List[(BigInt, Int)]): (Int, Int, Int, Int, Int, List[Int]) => List[(BigInt, Int)] = {

    def allJokerPairs(jokers: List[Int]): List[List[Int]] = {
      val jokerCards = jokers.sum
      if (jokerCards > 2) jokers :: allJokerPairs(nextBlocks(jokers))
      else if (jokerCards == 2) List(List(2), List(0))
      else List(List(0))
    }

    def areJokerBlocksAllowed(ranks: Int, suits: Int, pairs: Int, cards: Int): List[Int] => Boolean = jokers => {
      val jokerCards = jokers.sum
      val jokerPairs = if (jokerCards > 0) (jokers map (_ - 1) sum) else 0
      jokerPairs <= pairs && jokerCards <= cards && (pairs - jokerPairs) <= (BigDecimal((suits - 1) * math.min(ranks * suits, cards - jokerCards)) / BigDecimal(suits)).intValue
    }

    (ranks, suits, jokers, pairs, cards, blocks) =>

      if (jokers < 2)
        permutations(ranks, suits, jokers, pairs, cards, blocks)
      else {
        allJokerPairs(List(jokers)) filter areJokerBlocksAllowed(ranks, suits, pairs, cards) flatMap { j =>
          val jokersUsed = j.sum
          val jokerPairs = if (jokersUsed > 0) (j map (_ - 1) sum) else 0
          val jokerPerms = blockPermutations(1, jokers, if (jokersUsed > 0) 1 else 0, j)
          permutations(ranks, suits, jokers - jokersUsed, pairs - jokerPairs, cards - jokersUsed, blocks) map {
            case (x, y) => (x * jokerPerms, y + jokersUsed)
          }
        }
      }
  }

  /**
   * Probability of a pair being found in the first k cards
   */
  def pairProbability(k: Int, ranks: Int, suits: Int, jokers: Int = 0): BigDecimal = {
    val deck = ranks * suits + jokers
    def atLeastPPairs(p: Int): BigDecimal = {
      def probability(perm: (BigInt, Int)) = perm match {
        case (perms, cards) => BigDecimal(perms * (BigInt(k - cards + 1) to BigInt(k - p)).product) / BigDecimal((BigInt(deck - cards + 1) to BigInt(deck)).product)
      }
      val permutationsWithJokers = withJokers(permutations)
      permutationsWithJokers(ranks, suits, jokers, p, k, List(math.min(suits, k))) map probability sum
    }
    val maxPairs = math.floor((suits - 1) * k / suits).toInt
    if (maxPairs > 0) ((1 to maxPairs) map (p => BigDecimal(-1).pow(p + 1) * atLeastPPairs(p)) sum) else 0.0
  }

  /**
   * First pair probability
   */
  def firstPairProbability(k: Int, ranks: Int, suits: Int): BigDecimal = {
    if (k < 2)
      0.0
    else {
      def probabilityOfPairBefore(k: Int, ranks: Int, suits: Int, m: Int = 2): BigDecimal = {
        if (k <= m) 0.0
        else if (m == suits) pairProbability(k - suits, ranks - 1, suits)
        else pairProbability(k - m, ranks - 1, suits, suits - m) + (1.0 - probabilityOfPairBefore(k, ranks, suits, m + 1)) * BigDecimal(suits - m) / BigDecimal(ranks * suits - m)
      }
      (1.0 - probabilityOfPairBefore(k, ranks, suits)) * BigDecimal(suits - 1) / BigDecimal(ranks * suits - 1)
    }
  }

  /**
   * Mean location of first pair
   */
  def firstPairMeanLocation(ranks: Int, suits: Int): BigDecimal = { Random.shuffle(1 to (ranks * suits)).par map (k => { println(k); k * firstPairProbability(k, ranks, suits) }) sum }

  val Array(ranks, suits) = args map (_.toInt)
  val start = System.currentTimeMillis
  println(s"Mean location of first pair for ranks=$ranks, suits = $suits is: %s".format(firstPairMeanLocation(ranks, suits)))
  println("Time taken: %s (secs)".format((System.currentTimeMillis - start) / 1000.0))
}

trait Utils {
  def factorial(n: BigInt): BigInt = (BigInt(1) to n).product
  def binomialCoefficient(n: Int, k: Int) = (BigInt(n - k + 1) to n).product / (BigInt(1) to k).product
}
